pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -63,21 +96,52 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
-	followup, err := next.Execute(msg)
+
+	// trigger the event if the message type is registered (if it's not outbound)
+	eventsTriggered := false
+	if !msg.Outbound {
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
+		if err != nil {
+			return fmt.Errorf("send events failed: %w", err)
+		}
+	}
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// Register events on DID Exchange protocol. The events are triggered for incoming ConnectionRequest, ConnectionResponse
+// or ConnectionAck message types.
+func (s *Service) Register(ch chan<- event.DIDCommEvent) error {
+	s.event = append(s.event, ch)
```

Comment:
Is this thread safe?

Reply:
No. As per I remember during the last discussion, there were comments that registration happens when protocol is initiated. Hence locks are not necessary as such and having it will be an overhead during fetching the registered event to trigger it. Since we don't have that code now, I have added the mutex now.

Note - I had the locks earlier and removed it in previous patch.

Reply:
I don't know how registrations can happen when the protocol is initiated.

A client package is created "on the fly".